### PR TITLE
Automatic Hit Selection

### PR DESCRIPTION
**Pull request by labkey-nicka**
#### Rationale
This introduces the capability for automatic plate-based assay result hit selection based on filter criteria. Filter criteria are specified on assay result domain properties and are subsequently utilized to filter upon assay results when inserted/updated to determine if any should be marked as hits.

#### Related Pull Requests
- https://github.com/LabKey/labkey-api-js/pull/189
- https://github.com/LabKey/labkey-ui-components/pull/1649
- https://github.com/LabKey/labkey-ui-premium/pull/604
- https://github.com/LabKey/platform/pull/6123
- https://github.com/LabKey/limsModules/pull/941
- https://github.com/LabKey/commonAssays/pull/827
- https://github.com/LabKey/targetedms/pull/987
- https://github.com/LabKey/OConnorLabModules/pull/153

#### Changes
- Introduce `filterCriteria` to property descriptors and add CRUD support for them
- Add endpoint `assay-filterCriteriaColumns.api` to assist with retreival of column metadata for columns that are/to be generated
- Assay schema update to create the `assay.FilterCriteria` table
- Utilize Lombok to generated getters, setters, equals and hashCode for domain GWT POJOs
- Bump `@labkey/components`

### CODE REVIEW

File: assay/src/org/labkey/assay/TsvAssayProvider.java [hunk 1]
```diff
@@ -545,6 +565,204 @@ public boolean supportsSampleLookupsAsMaterialInputs()
...
+            return;
+
+        var table = AssayDbSchema.getInstance().getTableInfoFilterCriteria();
+
+        if (!toRemove.isEmpty())
+        {
+            var sql = new SQLFragment("DELETE FROM ").append(table)
+                    .append(" WHERE RowId ").appendInClause(toRemove, table.getSqlDialect());
+            new SqlExecutor(table.getSchema()).execute(sql);
+        }
```
Comment (labkey-klum) [inline assay/src/org/labkey/assay/TsvAssayProvider.java:738]: Are we transacting these changes or do we need to require a transaction?

File: assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java [hunk 2]
```diff
@@ -858,77 +863,227 @@ public OntologyManager.UpdateableTableImportHelper getImportHelper(
...
             if (!domainBaseProperties.contains(dp.getName()))
                 existingFields.put(dp.getName(), dp);
         });
 
-        for (GWTPropertyDescriptor prop : update.getFields())
+        return existingFields;
+    }
+
+    @Override
+    public Map<String, List<GWTPropertyDescriptor>> previewFilterCriteriaColumns(@NotNull ExpProtocol protocol, List<String> columnNames)
```
Comment (labkey-klum) [inline assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java:887]: The list of column names passed in assume that they are candidates for replicate stats (numeric measures). Does the list get filtered / validated on the client side, or is it possible to create replicate stats for a string field?

**Review by labkey-jeckels — COMMENTED** [review]
This looks fine to me from me in terms of Lombok usage.

File: assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java [hunk 3]
```diff
@@ -1170,6 +1317,104 @@ public void deleteReplicateStats(
...
+
+            if (domainProperty == null)
+            {
+                LOG.error("Automatic hit selection failed. Unable to resolve domain property from propertyId ({}).", criteria.propertyId());
+                return;
+            }
+
+            FieldKey fieldKey;
+            if (isReplicateProperty)
+                fieldKey = FieldKey.fromParts("Replicate", domainProperty.getName());
```
Comment (labkey-klum) [inline assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java]: Consider a string constant in the event we ever decide to change the FK to the replicate domain.

File: assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java [hunk 4]
```diff
@@ -1170,6 +1317,104 @@ public void deleteReplicateStats(
...
+        var filter = new SimpleFilter(FieldKey.fromParts("Well", "SampleId"), null, CompareType.NONBLANK);
+
+        // Filter out result rows that are excluded
+        {
+            PlateDataStateManager stateManager = PlateDataStateManager.get();
+            var exclusionStateRowIds = stateManager.getStates(container)
+                    .stream()
+                    .filter(state -> !stateManager.isOperationPermitted(state, PlateDataStateManager.DataOperation.hitSelection))
+                    .map(DataState::getRowId)
+                    .toList();
```
Comment (labkey-klum) [inline assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java]: You can probably just get the set of the `DataState` row IDs.

**Review by labkey-klum — APPROVED** [review]

File: assay/src/org/labkey/assay/TsvAssayProvider.java (hunk 1 above)
Comment (labkey-nicka) [reply assay/src/org/labkey/assay/TsvAssayProvider.java:738]: I've added an assertion.

File: assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java (hunk 3 above)
Comment (labkey-nicka) [reply assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java]: I've added an enumeration for columns that pertain to the `AssayResultsDomainKind`.

**Pull request merged**